<i>Modified from the file written by Ahsan Khan on behalf of Alberta Machine Intelligence Institute for the Al Pathways Partnership supported by Prairies Economic Development Canada</i>

---

**Important Note:**

Please do not alter any part of this notebook outside the designated text cells that are clearly marked with "*Start student input* ↓" and "*End student input ↑*". Changes made outside these specified areas could lead to incorrect evaluations of your work, potentially affecting your lab scores.

Ensure you complete all activities within these sections, which are indicated by labels like **[A1]**, **[A2]**, **[A3]**, ... Each activity is crucial for the successful completion of this lab. Additionally, please name your variables exactly as specified in the instructions (if specified) to ensure that your answers are correctly assessed.

---


# Lab 2: Decision Trees

Following up from the k-NN classifer, you will now be introduced to the Decision Tree (DT) classifier. DTs, like k-NN, are able to naturally handle non-linear multi-class data. The algorithm is not a distance based classifer like the K-NN classifer, instead it takes sequential  binary decisions in the internal nodes of the tree in order to arrive at a prediction (leaf). For this lab you will use the Breast Cancer Wisconsin (Diagnostic) dataset again.

In [ ]:
# Crucial data processing and analysis libraries
import numpy as np
import pandas as pd

# Loading the modules required to build and evaluate a DT classifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

# Loading the Breast Cancer Wisconsin (Diagnostic) dataset from sklearn
from sklearn.datasets import load_breast_cancer

##### Loading our data onto a dataframe the same way you encountered previously in lab 1.

In [ ]:
#loading data
breast_cancer = load_breast_cancer()

# There are three key parts to the dataset we care about
# The feature data, X
X = breast_cancer.data
display(X)

array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]])

In [ ]:
# The target classes, y
y = breast_cancer.target
display(y)

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

In [ ]:
# The feature names
display(breast_cancer.feature_names)

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

##### This is how the data looks like in a pandas dataframe

In [ ]:
df = pd.DataFrame(X, columns=breast_cancer['feature_names'])
df['class'] = y


print(f"Number of rows in the data: {df.shape[0]}")
df.head()

Number of rows in the data: 569


mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

##### Always a good idea to observe some statistics of our dataset to get an understanding of it.

In [ ]:
df.describe()

mean radius  mean texture  mean perimeter    mean area  \
count   569.000000    569.000000      569.000000   569.000000   
mean     14.127292     19.289649       91.969033   654.889104   
std       3.524049      4.301036       24.298981   351.914129   
min       6.981000      9.710000       43.790000   143.500000   
25%      11.700000     16.170000       75.170000   420.300000   
50%      13.370000     18.840000       86.240000   551.100000   
75%      15.780000     21.800000      104.100000   782.700000   
max      28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       mean symmetry  mean fractal dimension  ...  worst texture  \
count     569.000000              569.000000  ...     569.000000   
mean        0.181162                0.062798  ...      25.677223   
std         0.027414                0.007060  ...       6.146258   
min         0.106000                0.049960  ...      12.020000   
25%         0.161900                0.057700  ...      21.080000   
50%         0.179200                0.061540  ...      25.410000   
75%         0.195700                0.066120  ...      29.720000   
max         0.304000                0.097440  ...      49.540000   

       worst perimeter   worst area  worst smoothness  worst compactness  \
count       569.000000   569.000000        569.000000         569.000000   
mean        107.261213   880.583128          0.132369           0.254265   
std          33.602542   569.356993          0.022832           0.157336   
min          50.410000   185.200000          0.071170           0.027290   
25%          84.110000   515.300000          0.116600           0.147200   
50%          97.660000   686.500000          0.131300           0.211900   
75%         125.400000  1084.000000          0.146000           0.339100   
max         251.200000  4254.000000          0.222600           1.058000   

       worst concavity  worst concave points  worst symmetry  \
count       569.000000            569.000000      569.000000   
mean          0.272188              0.114606        0.290076   
std           0.208624              0.065732        0.061867   
min           0.000000              0.000000        0.156500   
25%           0.114500              0.064930        0.250400   
50%           0.226700              0.099930        0.282200   
75%           0.382900              0.161400        0.317900   
max           1.252000              0.291000        0.663800   

       worst fractal dimension       class  
count               569.000000  569.000000  
mean                  0.083946    0.627417  
std                   0.018061    0.483918  
min                   0.055040    0.000000  
25%                   0.071460    0.000000  
50%                   0.080040    1.000000  
75%                   0.092080    1.000000  
max                   0.207500    1.000000  

[8 rows x 31 columns]

# Lab activity: the decision tree classifer

##### **[A1]**  
Split your data into training and validation. Use ``X_train``, ``y_train``,``X_val`` and ``y_val`` as the assigned variables respectively.

*Start student input* ↓

In [ ]:
# Put your code here.
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2)

*End student input ↑*

##### **[A2]**
Instantiate a decision tree classifer called `dt`. For now, you do not need to set any hyperparameters or other class constructor arguments.

*Start student input* ↓

In [ ]:
# Put your code here.
dt = DecisionTreeClassifier()

*End student input ↑*

##### **[A3]**
Fit the decision tree classifier to your data

*Start student input* ↓

In [ ]:
# Put your code here.
dt.fit(X_train, y_train)

DecisionTreeClassifier()

##### **[A4]**
Predict on your validation data

*Start student input* ↓

In [ ]:
# Put your code here.
y_prediction=dt.predict(X_val)

*End student input ↑*

##### **[A5]**
Evaluate model performance using the `accuracy_score()` and  `classification_report()` functions (you can find the documentation for `classification_report` [here](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html).)

*Start student input* ↓

In [ ]:
# Put your code here.
# calculate the accuracy score
accuracyScore=accuracy_score(y_val, y_prediction)
print(accuracyScore)

# Generate classificaction report
classificationReport=classification_report(y_val, y_prediction)
print(classificationReport)

0.9122807017543859
              precision    recall  f1-score   support

           0       0.96      0.85      0.90        53
           1       0.88      0.97      0.92        61

    accuracy                           0.91       114
   macro avg       0.92      0.91      0.91       114
weighted avg       0.92      0.91      0.91       114



*End student input ↑*

Going back to your lab 1 notebook you may notice the accuracy for your first k-NN model was lower than the accuracy achieved in the first DT model here. Recall that you had to scale your data afterwards in order to achieve a decent accuracy score for the k-NN classifer.

##### **[A6]**
Normalize your dataset using the `StandardScaler()` function, normalizing both the `X_train` and `X_val` values from parameters set according to training data. Then, fit a new model `dt2` to scaled data. Finally evalaute the accuracy score.

*Start student input* ↓

In [ ]:
# Put your code here.
# import standardscaler for feature normalization
standardscaler = StandardScaler()
X_train_scaled = standardscaler.fit_transform(X_train)
X_val_scaled = standardscaler.transform(X_val)

#Instantiate a new Decision Tree classifier
dt2 = DecisionTreeClassifier()

dt2.fit(X_train_scaled, y_train)

y_prediction_scaled = dt2.predict(X_val_scaled)

# Compute the accuracy score of the model on the scaled data
accuracy_score_scaled = accuracy_score(y_val, y_prediction_scaled)
print(accuracy_score_scaled)


# Generate a classification report to analyze model performance
report_score_scaled = classification_report(y_val, y_prediction_scaled)
print(report_score_scaled)

0.9298245614035088
              precision    recall  f1-score   support

           0       0.96      0.89      0.92        53
           1       0.91      0.97      0.94        61

    accuracy                           0.93       114
   macro avg       0.93      0.93      0.93       114
weighted avg       0.93      0.93      0.93       114



*End student input ↑*

##### **[A7]**
Based on your results for the above two accuracy evaluations do you need to scale your data for a DT classifer? Explain.

*Start student input* ↓

Explanation: No, because a DT classifer doesnot used based computations , scaling is not required. The model functions well without normalizing, as evidenced by the little accuracy difference between before sclaing (91.2%) and after scaling (92.9%).Decisions tree are invarient to scaling since they divide data according to feature thresholds.

*End student input ↑*